util/ranger/detacher.go

```diff
@@ -364,6 +470,19 @@ func DetachCondAndBuildRangeForIndex(sctx sessionctx.Context, conditions []expre
 	for _, col := range cols {
 		newTpSlice = append(newTpSlice, newFieldType(col.RetType))
 	}
+	// In order to avoid exponential exploding, only expand once here.
+	if restConditions, dnfItems := expandDNF(conditions, cols); restConditions != nil && dnfItems != nil {
+		var rebuildItems []expression.Expression
+		var newConditions []expression.Expression
+		var rebuildConditions []expression.Expression
+		for _, item := range dnfItems {
+			tempConditions := make([]expression.Expression, len(restConditions))
+			copy(tempConditions, restConditions)
+			newConditions = append(tempConditions, item)
+			rebuildItems = append(rebuildItems, expression.ComposeCNFCondition(sctx, newConditions...))
+		}
+		conditions = append(rebuildConditions, expression.ComposeDNFCondition(sctx, rebuildItems...))
+	}
```

Comment:
How about putting this logic into `expandDNF`, which may make this code clearer.
```
conditions = expandDNF(conditions, cols)
```

Reply:
OK